backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,126 @@
+            return new RegisterUserResponse(null, null, "The username" +
+                    " is already occupied. If that is previously yours, consider logging in instead?");
+        }
+
+        // Nothing goes wrong
+        return null;
+    }
+
+    /**
+     * A method for future different types of {@link User} that might need to validate
+     * their own defined attributes (i.e. beyond the uniqueness check in username and email address)
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     *
+     * @return true if assumptions for registering that type of {@link User} holds; false otherwise
+     */
+    protected boolean isValidated(T boundary){
+        /* I think this should be overridden in case to add some further checks for future user types.
+        isValidRegistering is "good enough" for clients and provider for now*/
+        return isValidRegistering(boundary) == null;
+    }
+
+    /**
+     * A method to bundle up {@link UserInfo} based on their given
+     * {@link com.javahelp.model.user.UserType}
+     */
+    protected U bundleInfo(T boundary){
+        return null; // UserInfo is designed as an abstract class so nothing can be initialized here
+    }
+
+    /**
+     * Process registering a {@link User}
+     * Note: This is only called after all registration tests are passed.
+     *
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     * @param user a newly generated {@link User}
+     *
+     * @return a {@link RegisterUserResponse} that contains all result info.
+     */
+    protected RegisterUserResponse registering(T boundary, User user) {
+        // Make a new user account as well as its password
+        User newUser = userStore.create(user, boundary.getPassword());
+        UserPassword password = userStore.readPassword(newUser.getStringID());
+
+        return new RegisterUserResponse(newUser, password, null);
+    }
+
+    /**
+     * To create a new user entity for this user based on their info:
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to get info from the
+     *                 register View Model.
+     *
+     * @return a new {@link RegisterUserResponse} entity that comprises of {@link User}
+     * {@link UserPassword} and {@link com.javahelp.model.token.Token}
+     */
+    public RegisterUserResponse create(T boundary){
+        // Check the assumptions
+        boolean result = this.isValidated(boundary);
+
+        if (!result){
+            // Early return for failure. Call isValidRegistering to get a full response model
+            return this.isValidRegistering(boundary);
+        }
+
+        // After all checks passed, time to make a new user
+        User user = register(boundary);
+
+        return registering(boundary, user);
+    }
```

Comment:
perhaps this method should be called `register` or something. Also, the structure of calling `isValidated` then if it returns false, calling `isValidRegistering` should  be changed. `isValidRegistering` should be called. If it comes back non null, its output should be returned. The same should then be done with `isValidated`.

Reply:
Additionally, you don't need to qualify every use of a method or class field with `this`. It might be best practice to only do so when it remove ambiguity and is necessary for the code. Most modern IDEs display local and class fields differently so it does not really add to readability.